### PR TITLE
Fix #16778 - use previous type and not declared type

### LINKED ISSUE #16778: Flow control analysis is not performed on discriminatable types flowed from a type guard
**TypeScript Version:**  2.4.0 / nightly (2.5.0-dev.201xxxxx)

**Code**

```ts
enum Kind {
    A,
    B,
}

interface Base {
    kind: Kind
}

interface A extends Base { 
    kind: Kind.A;
    yar: any;
}

interface B extends Base {
    kind: Kind.B;
    gar: any;
}

type Both = A | B;
function isBoth(x: Base): x is Both {
    return true;
}

let foo: Base = undefined;
if (isBoth(foo)) {
    switch (foo.kind) {
        case Kind.A:
            foo; // <-- type of foo here
    }
}
```

**Expected behavior:**
`foo` to be of type `A` at the indicated line.
**Actual behavior:**
`foo` is of type `Both`.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #16778. `isMatchingReferenceDiscriminant` needed to use the previous type witnessed for the reference in question rather than its declared type to check if it was discriminable.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11627,11 +11627,11 @@ namespace ts {
...
             }
 
-            function isMatchingReferenceDiscriminant(expr: Expression) {
+            function isMatchingReferenceDiscriminant(expr: Expression, prevType: Type) {
                 return expr.kind === SyntaxKind.PropertyAccessExpression &&
-                    declaredType.flags & TypeFlags.Union &&
+                    prevType.flags & TypeFlags.Union &&
                     isMatchingReference(reference, (<PropertyAccessExpression>expr).expression) &&
-                    isDiscriminantProperty(declaredType, (<PropertyAccessExpression>expr).name.escapedText);
+                    isDiscriminantProperty(prevType, (<PropertyAccessExpression>expr).name.escapedText);
```
Comment (sandersn) [inline src/compiler/checker.ts]: prevType has a specific name elsewhere in the control flow code. I'll go find out what it is.

Comment (sandersn) [inline src/compiler/checker.ts]: Aha! "computed type"

**Comment by sandersn** [comment]
Note: I would have been a lot more enthusiastic about this fix if you had used our real Node code from the compiler in the test.

**Comment by weswigham** [comment]
> Note: I would have been a lot more enthusiastic about this fix if you had used our real Node code from the compiler in the test.

That would have been significantly greater than the minimum required code to repro and examine the bug. XD

**Pull request merged**